src/main.rs

```diff
@@ -0,0 +1,107 @@
+use miden_crypto::{
+    hash::rpo::RpoDigest,
+    merkle::{MerkleError, NodeIndex},
+    Felt, Word, ONE,
+    {hash::rpo::Rpo256, merkle::TieredSmt},
+};
+use std::time::Instant;
+
+fn main() {
+    construction(1000).unwrap();
+    construction(1_000_000).unwrap();
+
+    insertion(1000).unwrap();
+    insertion(1_000_000).unwrap();
+
+    proof_generation(1000).unwrap();
+    proof_generation(1_000_000).unwrap();
+}
+
+pub fn construction(size: u64) -> Result<(), MerkleError> {
+    let mut leaves = Vec::new();
+    for i in 0..size {
+        let key = Rpo256::hash(&rand_utils::rand_value::<u64>().to_be_bytes());
+        let value = [ONE, ONE, ONE, Felt::new(i)];
+        leaves.push((key, value));
+    }
+
+    let now = Instant::now();
+    let tree = TieredSmt::with_leaves(leaves)?;
+    let elapsed = now.elapsed();
+    println!("Elapsed: {:.2?}", elapsed);
+
+    get_nodes_num_by_tier(&tree);
+    Ok(())
+}
+
+pub fn insertion(size: u64) -> Result<(), MerkleError> {
+    let mut leaves = Vec::new();
+    for i in 0..size {
+        let key = Rpo256::hash(&rand_utils::rand_value::<u64>().to_be_bytes());
+        let value = [ONE, ONE, ONE, Felt::new(i)];
+        leaves.push((key, value));
+    }
+
+    let mut tree = TieredSmt::with_leaves(leaves)?;
+    let test_key = Rpo256::hash(&rand_utils::rand_value::<u64>().to_be_bytes());
+    let test_value = [ONE, ONE, ONE, Felt::new(size + 1)];
+
+    let now = Instant::now();
+    tree.insert(test_key, test_value);
+    let elapsed = now.elapsed();
+    println!("Elapsed: {:.2?}", elapsed);
+
+    get_nodes_num_by_tier(&tree);
+    Ok(())
+}
+
+pub fn proof_generation(size: u64) -> Result<(), MerkleError> {
```

Comment:
I would pass in the tree into this function as a parameter instead of building it here.